Tasks/CargoAuthenticateV0/cargoauthenticatemain.ts

```diff
@@ -0,0 +1,76 @@
+import * as path from 'path';
+import * as tl from 'azure-pipelines-task-lib/task';
+import * as toml from 'toml';
+import * as fs from 'fs';
+import * as constants from './constants';
+import * as util from 'azure-pipelines-tasks-packaging-common/util';
+import * as pkgLocationUtils from 'azure-pipelines-tasks-packaging-common/locationUtilities';
+import * as url from 'url';
+
+async function main(): Promise<void> {
+    tl.setResourcePath(path.join(__dirname, 'task.json'));
+
+    try {
+        let configtoml = tl.getInput(constants.CargoAuthenticateTaskInput.WorkingFile);
+        if (!tl.exist(configtoml)) {
+            throw new Error(tl.loc('ConfigTomlDoesNotExist', configtoml));
+        }
+        else {
+            console.log(tl.loc("AuthenticatingThisConfigToml", configtoml));
+        }
+        
+        // These two formats will work
+        // [registries]
+        // zhentan-test = { index = "sparse+https://pkgs.dev.azure.com/codesharing-SU0/zhentan-test/_packaging/zhentan-test/Cargo/index/" }
+        // [registries.zhentan-test1]
+        // index = "sparse+https://pkgs.dev.azure.com/codesharing-SU0/zhentan-test/_packaging/zhentan-test/Cargo/index/"
+        let configtomlFile = fs.readFileSync(configtoml, 'utf8');
+        var result = toml.parse(configtomlFile);
+        if (!result.registries)
+        {
+            throw new Error("config.toml must contains registries field");
+        }
+
+        let packagingLocation: pkgLocationUtils.PackagingLocation;
+        try {
+            packagingLocation = await pkgLocationUtils.getPackagingUris(pkgLocationUtils.ProtocolType.Npm);
+        } catch (error) {
+            tl.debug('Unable to get packaging URIs');
+            util.logError(error);
+            throw error;
+        }
+
+        const collectionHosts = packagingLocation.PackagingUris.map((pkgUrl: string) => {
+            const parsedUrl = url.parse(pkgUrl);
+            if (parsedUrl && parsedUrl.host) {
+                return parsedUrl.host.toLowerCase();
+            }
+            return undefined;
+        });
+
+        const token = `Bearer ${tl.getVariable('System.AccessToken')}`;
+        for (let registry of Object.keys(result.registries)) {
+            const registryUrl = url.parse(result.registries[registry].index);
+            if(registryUrl && registryUrl.host && collectionHosts.indexOf(registryUrl.host.toLowerCase()) >= 0) {
+                const tokenName = `CARGO_REGISTRIES_${registry.toLocaleUpperCase().replace("-", "_")}_TOKEN`;
+                tl.debug(tl.loc('AddingAuthRegistry', registry, tokenName));
+                tl.setVariable(tokenName, token);
+            }   
+        }
+
+        //External endpoints
+    }
+
+    catch(error) {
+        tl.error(error);
+        tl.setResult(tl.TaskResult.Failed, tl.loc("FailedToAddAuthentication"));
+        return;
+    }
+
+    finally {
+        // What telemetry we can emit?
```

Comment:
Emit standard task telemetry at least, you can skip adding custom props if you want

Reply:
We already have some at line 56, not sure if we need anything else since we are just getting build identity and set it to environment variable.